metadata-ingestion/src/datahub/utilities/sqlalchemy_query_combiner.py

```diff
@@ -0,0 +1,345 @@
+            return self._result[0]
+        elif self._result:
+            raise MultipleResultsFound("Multiple rows returned for one()")
+        else:
+            raise NoResultFound("No rows returned for one()")
+
+    def one_or_none(self) -> Optional[Any]:
+        if len(self._result) == 1:
+            return self._result[0]
+        elif self._result:
+            raise MultipleResultsFound("Multiple rows returned for one_or_none()")
+        else:
+            return None
+
+    def scalar(self) -> Any:
+        if len(self._result) == 1:
+            row = self._result[0]
+            return row[0]
+        elif self._result:
+            raise MultipleResultsFound(
+                "Multiple rows were found when exactly one was required"
+            )
+        return None
+
+    def update(self) -> None:
+        # No-op.
+        pass
+
+    def close(self) -> None:
+        # No-op.
+        pass
+
+    all = fetchall
+    fetchone = one
+
+
+@dataclasses.dataclass
+class _QueryFuture:
+    conn: Connection
+    query: sqlalchemy.sql.Select
+    multiparams: Any
+    params: Any
+
+    done: bool = False
+    res: Optional[_ResultProxyFake] = None
+    exc: Optional[Exception] = None
+
+
+@dataclasses.dataclass
+class SQLAlchemyQueryCombiner:
+    """
+    This class adds support for dynamically combining multiple SQL queries into
+    a single query. Specifically, it can combine queries which each return a
+    single row. It uses greenlets to manage the execution lifecycle of the queries.
+    """
+
+    enabled: bool
+    catch_exceptions: bool
+    is_single_row_query_method: Callable[[Any], bool]
+    serial_execution_fallback_enabled: bool
+
+    # There will be one main greenlet per thread. As such, queries will be
+    # queued according to the main greenlet's thread ID. We also keep track
+    # of the greenlets we spawn for bookkeeping purposes.
+    _thread_unsafe_operation_lock: threading.Lock = dataclasses.field(
+        default_factory=lambda: threading.Lock()
+    )
+    _queries_by_thread: Dict[
+        greenlet.greenlet, Dict[str, _QueryFuture]
+    ] = dataclasses.field(default_factory=dict)
+    _greenlets_by_thread: Dict[
+        greenlet.greenlet, Set[greenlet.greenlet]
+    ] = dataclasses.field(default_factory=lambda: collections.defaultdict(set))
+
+    def _get_main_greenlet(self) -> greenlet.greenlet:
+        let = greenlet.getcurrent()
+        while let.parent is not None:
+            let = let.parent
+        return let
+
+    def _get_queue(self, main_greenlet: greenlet.greenlet) -> Dict[str, _QueryFuture]:
+        assert main_greenlet.parent is None
+
+        with self._thread_unsafe_operation_lock:
+            return self._queries_by_thread.setdefault(main_greenlet, {})
+
+    def _get_greenlet_pool(
+        self, main_greenlet: greenlet.greenlet
+    ) -> Set[greenlet.greenlet]:
+        assert main_greenlet.parent is None
+
+        with self._thread_unsafe_operation_lock:
```

Comment:
why is the same lock protecting read from `self._greenlets_by_thread` and setting to `self._queries_by_thread` on line 148?

Reply:
Good catch - will change!